src/sympc/protocol/falcon/falcon.py

```diff
@@ -451,3 +454,87 @@ def select_shares(x: MPCTensor, y: MPCTensor, b: MPCTensor) -> MPCTensor:
         z = x + (d * (y - x))
 
         return z
+
+    @staticmethod
+    def _random_prime_group(
+        session: Session, shape: Union[torch.Size, tuple]
+    ) -> MPCTensor:
+        """Computes shares of random number in Zp*.Zp* is the multiplicative group mod p.
+
+        Args:
+            session (Session): session to generate random shares for.
+            shape (Union[torch.Size,tuple]): shape of the random share to generate.
+
+        Returns:
+            share (MPCTensor): Retuns shares of random number in group Zp*.
+
+        Zp* = {1,2..,p-1},where p is a prime number.
+        We use Euler's Theorum for verifying that random share is not zero.
+        It states that:
+        For a general modulus n
+        a^phi(n) = 1(mod n), if a is co prime to n.
+        In our case n=p(prime number), phi(p) = p-1
+        phi(n) = Euler totient function.
+        """
+        while True:
+            ptr_list: List[ReplicatedSharedTensor] = []
+            for session_ptr in session.session_ptrs:
+                ptr = session_ptr.prrs_generate_random_share(
+                    shape=(), ring_size=str(PRIME_NUMBER)
+                ).resolve_pointer_type()
+                ptr = ptr.repeat(shape)
+                ptr_list.append(ptr)
+
+            m = MPCTensor(shares=ptr_list, session=session, shape=shape)
+
+            m_euler = m ** (PRIME_NUMBER - 1)
+
+            if (m_euler.reconstruct(decode=False) == 1).all():
+                return m
+
+    @staticmethod
+    def private_compare(x: List[MPCTensor], r: torch.Tensor) -> MPCTensor:
+        """Falcon Private Compare functionality which computes(x>r).
+
+        Args:
+            x (List[MPCTensor]) : shares of bits of x in Zp.
+            r (torch.Tensor) : Public integer r.
+
+        Returns:
+            result (MPCTensor): Returns shares of bits of the operation.
+
+        (if (x>=r) returns 1 else returns 0)
+        """
+        shape = x[0].shape
+        session = x[0].session
+        ptr_list: List = []
+        for session_ptr in session.session_ptrs:
+            ptr_list.append(
+                session_ptr.prrs_generate_random_share(shape=shape, ring_size=str(2))
+            )
+
+        beta_2 = MPCTensor(
+            shares=ptr_list, session=session, shape=shape
+        )  # shares of random bit
+        beta_p = ABY3.bit_injection(
+            beta_2, session, PRIME_NUMBER
+        )  # shares of random bit in Zp.
+        m = Falcon._random_prime_group(session, shape)
+
+        u = [0 for i in range(len(x))]
+        w = [0 for i in range(len(x))]
+        c = [0 for i in range(len(x))]
+        for i in range(len(x) - 1, -1, -1):
+            r_i = r >> i & 1  # bit at ith position
+            u[i] = (1 - 2 * beta_p) * (x[i] - r_i)
+            w[i] = x[i] ^ r_i
+            c[i] = u[i] + 1 + sum(w[i + 1 :])
```

Comment:
Q: Isn't a bug here? when `i == len(x) - 1` we do `w[len(x) :])` but `w` can be indexed from `0 to len(x) - 1`

Reply:
Good question, george, I will effectively wanted it to return 0 ,when `i==len(x)-1` , when we slice `w[i+1:]` ,it returns an empty list, when we use sum on top of empty,it returns zero,which fit the scenario.

Reply:
Ahh...Nice - interesting behavior for python - I will recheck this logic today.

Reply:
MAYBE: Can't we keep `w` as a single number?
```
w = 0
for i in range(len(x) - 1, -1, -1):
            r_i = r >> i & 1  # bit at ith position
            u[i] = (1 - 2 * beta_p) * (x[i] - r_i)
            w += x[i] ^ r_i
            c[i] = u[i] + 1 + w
```

In this case, you won't have to always do the `sum`.

Reply:
You mean something like also sanity check george, like when
<code>if i==len(x)-1 <br>
w=0
</code>

Reply:
Crct George,  running sum is inefficient in the context of MPCTensor, I would keep and current variable at end and change it

Reply:
Modified.